pkg/plugins/secret/secret_manager/secret_manager_test.go

```diff
@@ -0,0 +1,310 @@
+
+					By("Returning the service provided version id")
+					Expect(response.SecretVersion.Version).To(Equal("1"))
+				})
+			})
+
+			When("Putting a Secret to a non-existent secret", func() {
+				crtl := gomock.NewController(GinkgoT())
+				mockSecretClient := mocks.NewMockSecretManagerClient(crtl)
+				secretPlugin := &secretManagerSecretService{
+					client:    mockSecretClient,
+					projectId: "my-project",
+				}
+
+				It("Should successfully store a secret", func() {
+					defer crtl.Finish()
+
+					//Mocking expects
+					By("Calling SecretManagerService.GetSecret")
+					mockSecretClient.EXPECT().GetSecret(
+						gomock.Any(),
+						&secretmanagerpb.GetSecretRequest{
+							Name: "projects/my-project/secrets/Test",
+						},
+					).Return(nil, status.Error(codes.NotFound, "secret not found")).Times(1)
+
+					By("Calling SecretManagerService.CreateSecret with the expected payload")
+					mockSecretClient.EXPECT().CreateSecret(
+						gomock.Any(),
+						&secretmanagerpb.CreateSecretRequest{
+							Parent:   "projects/my-project",
+							SecretId: "Test",
+							Secret: &secretmanagerpb.Secret{
+								Replication: &secretmanagerpb.Replication{
+									Replication: &secretmanagerpb.Replication_Automatic_{
+										Automatic: &secretmanagerpb.Replication_Automatic{},
+									},
+								},
+							},
+						},
+					).Return(&secretmanagerpb.Secret{
+						Name: "projects/my-project/secrets/Test",
+					}, nil).Times(1)
+
+					By("Calling SecretManagerService.AddSecretVersion")
+					mockSecretClient.EXPECT().AddSecretVersion(
+						gomock.Any(),
+						&secretmanagerpb.AddSecretVersionRequest{
+							Parent: "projects/my-project/secrets/Test",
+							Payload: &secretmanagerpb.SecretPayload{
+								Data: []byte("Super Secret Message"),
+							},
+						},
+					).Return(&secretmanagerpb.SecretVersion{
+						Name: "/projects/my-project/Test/versions/1",
+					}, nil).Times(1)
+
+					response, err := secretPlugin.Put(&testSecret, testSecretVal)
+					By("Not returning an error")
+					Expect(err).ShouldNot(HaveOccurred())
+
+					By("Returning a response with a version id")
+					Expect(response.SecretVersion.Version).To(Equal("1"))
+				})
+			})
+
+			When("Putting a nil secret", func() {
+				secretPlugin := &secretManagerSecretService{
+					projectId: "my-project",
+				}
+
+				It("Should return an error", func() {
+					_, err := secretPlugin.Put(nil, testSecretVal)
+					Expect(err).Should(HaveOccurred())
+					Expect(err.Error()).To(Equal("secret must not be nil"))
+				})
+			})
+
+			When("Putting a secret with an empty name", func() {
+				secretPlugin := &secretManagerSecretService{
+					projectId: "my-project",
+				}
+
+				It("Should return an error", func() {
+					var emptySecretName = &secret.Secret{}
+					_, err := secretPlugin.Put(emptySecretName, testSecretVal)
+
+					Expect(err).Should(HaveOccurred())
+					Expect(err.Error()).To(Equal("provide non-blank secret name"))
+				})
+			})
+
```

Comment:
Could we add some dodgy secret name validation as well.

Reply:
new issue for this, as we should add this in the adapter layer.